agents/check_mk_agent.linux

```diff
@@ -415,6 +415,69 @@ run_real_time_checks() {
     done
 }
 
+# Function to pull out NTP metrics
+section_ntp() {
+    # ntpq helper function
+    get_ntpq() {
+        inpath ntpq || return 1
+        [ "${1}" = "--header" ] && echo '<<<ntp>>>'
+        ntpq -np | sed -e 1,2d -e 's/^\\(.\\)/\\1 /' -e 's/^ /%/' || true
+    }
+
+    # First we check for systemd, and then attempt the methods that we know about
+    # Our goal here is to avoid multiple outputs e.g. timedatectl + ntpq
+    if inpath systemctl; then
+        # Attempt to check time synchronization with systemd-timesyncd
+        if inpath timedatectl; then
+            # This timedatectl arg seems to be available from systemd version >=239
+            if timedatectl timesync-status >/dev/null 2>&1; then
+                echo "<<<timesyncd>>>"
+                timedatectl timesync-status
+                get_file_mtime /var/lib/systemd/timesync/clock | awk '{print "[[["$1"]]]"}'
+
+                # Return to get out of 'section_ntp()' with no further processing
+                return
+            fi
+        fi
+
+        # Otherwise, try to identify whether ntp/ntpd/openntpd is active, if so, run ntpq
+        # shellcheck disable=SC2016
+        if [ "$(systemctl | awk '/ntp.service|ntpd.service/{print $3; exit}')" = "active" ]; then
+            # remove heading, make first column space separated
+            run_cached -s ntp 30 "waitmax 5 get_ntpq"
+            return
+        fi
+    fi
+
+    # Next, we try to check time synchronization with Chrony
+    # Force successful exit code. Otherwise section will be missing if daemon not running
+    #
+    # The "| cat" has been added for some kind of regression in RedHat 7.5. The
+    # SELinux rules shipped with that release were denying the chronyc call without cat.
+    if inpath chronyc; then
```

Comment:
This will not execute the chrony section if the timesync information is successfully retrieved. Could you open a separate request for the chrony section and leave this out from this commit?

Reply:
Hi Anthony,
I'd prefer not to deal with `chrony` separately unless there is a really good justification.  The point of this PR is to get the ongoing issues with _all_ of the time information gathering - _which includes `chrony`_ - squared away.

So, to clarify the requirements, you want to `chrony` to run:

* If it's a systemd based host that _can_ run `timedatectl timesync-status`
* If it's a systemd based host that _can't_ run `timedatectl timesync-status` but _can_ run `ntpq`
* If it's a systemd based host that _can't_ run `timedatectl timesync-status` and _can't_ run `ntpq`
* If it's not a systemd based host yet somehow still has `chrony`

Is that correct?  Or is there some other list of requirements?

Reply:
Hi Rawiri, 

I had a slightly longer discussion internally as we were already planning some bigger changes to our agent.
In order to prepare for this future work, I would like to separate timesyncd, ntp and chrony in their own respective sections.  Maintain the current relationship by bundling timesyncd and ntp verification and run chrony independently. The structure could look as follows: 

- section_timesyncd with a return code for a succesful execution
- section_ntp: depending on the return code of section_timesyncd, this section gets called (your two ntp verifications including the systemd are included here)
- section_chrony: gets always executed for now regardless of section_timesyncd & section_ntp (a double time synchronisation check is deemed acceptable for now)

This is slightly different to your current iteration (and also your last remark) but your proposed verification mechanisms would be still very valuable. We will be working towards a more complete unification of the time sychronisation processes (similar to what you suggested) either in the agent directly or through a more general approach related to the agent soon.

Best, 

Anthony

Reply:
Hi Anthony,
Have a look at 2845daa and see if that's any closer to what you're after.

It may also help me to have any visibility of what bigger changes are planned for the agent.

Cheers

Rawiri

Reply:
Hi Rawiri, 

Yes. Is it possible to extract all the ntp calls into their own section? Even if it means that the systemd verification may happen twice? Otherwise, the section calls look good to me.

Reply:
Hi Anthony,
I've been trying to avoid that.  With the merged agent script, I've tried to bring in more of a cleaner separation between environment and local variable "scopes".  I'm fairly sure that in doing so, I cleaned everything out of the script's scope.  Using a variable to change the behaviour of `section_ntp()` undoes that a bit and just *feels* a bit messy.

But I've given it a shot anyway, see 8cd85ef

Rawiri

Reply:
OK, I think I have everything in order now.  Please check 8af56c4

Reply:
Hi Rawiri,

sorry for the late reply. I modified some of your changes (to accommodate our checks) and merged most of your suggestions this morning. Please have a look at it. 

Thanks

Anthony

Reply:
Hey Anthony,
at first glance it looks like you've made some sane changes, I like it :)  The only tweak I'd suggest is removing the handling for $timesync_rc from get_ntpq(), as it's redundant now.

And maybe add a TO-DO note under "# Try to determine status via /etc/init.d" saying something like "TO-DO: Check how to do this under OpenRC"

As an aside: Is there someone (or a generic email address) that I can contact off-github regarding future plans for the *nix agent?  I'd like to continue contributing, and have some ideas for a major restructure, but I'm doing a lot of flying blind here...

Rawiri

Reply:
Hi Rawiri,

I am glad that you liked my changes since the commit was authored by you. I changed the timesync_rc aspect as well. I will put your To-Do note in our system. 

Regarding your aside: We always take a look at the pull requests here. Therefore, regarding smaller suggestions you are welcome to post them here. Otherwise, for now I would suggest using our official mail info@checkmk.com for your suggestions. 

Best, 

Anthony